File: 1.arrays/binary-search/binary_search.cs
using System;
namespace BinarySearch
{
  class Propgram
  {
    static void Main(string[] args)
    {
      int[] arr = new int[] { 3, 5, 12, 56, 92, 123, 156, 190, 201, 222 };
      int key = 12;
      Console.WriteLine(binarySearch(arr, key));
    }
    static int binarySearch(int[] arr, int key)
    {
      int lower = 0;
      int upper = arr.Length;
      int indexMid = 0;
      int mid = 0;
      while (lower < upper)
      {
        indexMid = lower + (upper - lower) / 2;
        mid = arr[indexMid];
        if (mid == key)
        {
          return indexMid;
        }
        else if (mid < key)
        {
          lower = indexMid + 1;
        }
        else
        {
          upper = indexMid - 1;
        }
      }
      return -1;
    }
  }
}

File: 1.arrays/binary-search/binary_search.rb
def binary_search(arr, key)
  lower = 0
  upper = arr.length
  while lower<upper do
    indexMid = (lower + (upper -  lower)/2).round(0)
    mid = arr[indexMid]
    if key == mid
      return indexMid
    elsif mid > key
      upper = indexMid -1
    else
      lower = indexMid + 1
    end
  end
  return -1
end

arr = [3, 5, 12, 56, 92, 123, 156, 190, 201, 222]
key = 12
result = binary_search(arr, key)
puts result


# Complexity
# Worst case time complexity: O(log N)
# Average case time complexity: O(log N)
# Best case time complexity: O(1)
# Space complexity: O(1)